examples/rcn.py

```diff
@@ -0,0 +1,392 @@
+        A configuration matrix (shape n X 2) where n is the number of valid configurations.
+    """
+
+    configs = []
+    for i, (r1, c1) in enumerate(np.array(np.unravel_index(np.arange(M), (25, 25))).T):
+        r2_min = max(r1 - r, 0)
+        r2_max = min(r1 + r, 2 * hps)
+        c2_min = max(c1 - r, 0)
+        c2_max = min(c1 + r, 2 * vps)
+        j = np.ravel_multi_index(
+            tuple(np.mgrid[r2_min : r2_max + 1, c2_min : c2_max + 1]), (25, 25)
+        ).ravel()
+        configs.append(np.stack([np.full(j.shape, fill_value=i), j], axis=1))
+
+    return np.concatenate(configs)
+
+
+max_perturb_radius = 25
+valid_configs_list = [valid_configs(r) for r in range(max_perturb_radius)]
+
+# %% [markdown]
+# ### 4.2.2 Make the factor graph
+
+# %%
+start = end
+fg = graph.FactorGraph(variables=variables_all_models)
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
+    for e in edge:
+        i1, i2, r = e
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            valid_configs_list[r],
+        )
+
+end = time.time()
+print(f"Creating factors took {end-start:.3f} seconds.")
+
+
+# %% [markdown]
+# # 5. Run inference
+
+# %% [markdown]
+# ## 5.1 Helper functions to initialize the evidence for a given image
+
+# %%
+def get_bu_msg(img: np.ndarray) -> np.ndarray:
+    """Computes the bottom-up messages given a test image.
+
+    Args:
+        img: The rgb image to compute bottom up messages on (H x W x 3).
+
+    Returns:
+        An array of shape [16 x H x W] denoting the presence or absence of an oriented line segment at a particular location.
+            The elements of this array belong to the set {+1, -1}.
+    """
+
+    num_orients = 16
+    brightness_diff_threshold = 40.0
+
+    filtered = np.zeros((filters.shape[0],) + img.shape, dtype=np.float32)
+    for i in range(filters.shape[0]):
+        kern = filters[i, :, :]
+        filtered[i] = fftconvolve(img, kern, mode="same")
+
+    localized = np.zeros_like(filtered)
+    cross_orient_max = filtered.max(0)
+    filtered[filtered < 0] = 0
+    for i, (layer, suppress_mask) in enumerate(zip(filtered, suppression_masks)):
+        competitor_maxs = maximum_filter(layer, footprint=suppress_mask, mode="nearest")
+        localized[i] = competitor_maxs <= layer
+    localized[cross_orient_max > filtered] = 0
+
+    # Threshold and binarize
+    localized *= (filtered / brightness_diff_threshold).clip(0, 1)
+    localized[localized < 1] = 0
+
+    pooled_channel_weights = [(0, 1), (-1, 1), (1, 1)]
+    pooled_channels = [-np.ones_like(sf) for sf in localized]
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset, factor in pooled_channel_weights:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            if factor != 1:
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
+
+    bu_msg = np.array(pooled_channels)
+    bu_msg[bu_msg == 0] = -1
```

Comment:
Add comment explaining what you are doing here and why.

Reply:
Added explanation.